Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,726 @@
+        }
+
+        switch (maybe_result.value()) {
+        case RecursionDecision::Recurse: {
+            auto maybe_has_moved = iterator.next();
+            if (maybe_has_moved.is_error()) {
+                return maybe_has_moved.error();
+            }
+
+            if (!maybe_has_moved.value()) {
+                // If next() hasn't moved then we have read through all the
+                // directories, and can exit.
+                return KSuccess;
+            }
+
+            continue;
+        }
+        case RecursionDecision::Continue: {
+            while (!iterator.done()) {
+                if (iterator.skip())
+                    break;
+                if (!iterator.go_up())
+                    return KSuccess;
+            }
+
+            continue;
+        }
+        case RecursionDecision::Break:
+            return KSuccess;
+        }
+    }
+
+    return KSuccess;
+}
+
+KResultOr<NonnullRefPtr<ISO9660FS::DirectoryEntry>> ISO9660FS::directory_entry_for_record(Badge<ISO9660DirectoryIterator>, ISO::DirectoryRecordHeader const* record)
+{
+    u32 extent_location = LittleEndian { record->extent_location.little };
+    u32 data_length = LittleEndian { record->data_length.little };
+
+    auto key = calculate_directory_entry_cache_key(record);
+    auto it = m_directory_entry_cache.find(key);
+    if (it != m_directory_entry_cache.end()) {
+        dbgln_if(ISO9660_DEBUG, "Cache hit for dirent @ {}", extent_location);
+        return it->value;
+    }
+    dbgln_if(ISO9660_DEBUG, "Cache miss for dirent @ {} :^(", extent_location);
+
+    if (m_directory_entry_cache.size() == max_cached_directory_entries) {
+        // FIXME: A smarter algorithm would probably be nicer.
+        m_directory_entry_cache.remove(m_directory_entry_cache.begin());
+    }
+
+    if (!(data_length % logical_block_size() == 0)) {
+        dbgln_if(ISO9660_DEBUG, "Found a directory with non-logical block size aligned data length!");
+        return EIO;
+    }
+
+    auto blocks = KBuffer::try_create_with_size(data_length, Region::Access::Read | Region::Access::Write, "ISO9660FS: Directory traversal buffer");
+    if (!blocks) {
+        return ENOMEM;
+    }
+
+    auto blocks_buffer = UserOrKernelBuffer::for_kernel_buffer(blocks->data());
+    auto did_read = raw_read_blocks(BlockBasedFileSystem::BlockIndex { extent_location }, data_length / logical_block_size(), blocks_buffer);
+    if (!did_read) {
+        return EIO;
+    }
+
+    auto maybe_entry = DirectoryEntry::create(extent_location, data_length, move(blocks));
+    if (maybe_entry.is_error()) {
+        return maybe_entry.error();
+    }
+    m_directory_entry_cache.set(key, maybe_entry.value());
+
+    dbgln_if(ISO9660_DEBUG, "Cached dirent @ {}", extent_location);
+    return maybe_entry.release_value();
+}
+
+u32 ISO9660FS::calculate_directory_entry_cache_key(ISO::DirectoryRecordHeader const* record)
+{
+    return LittleEndian { record->extent_location.little };
+}
```

Comment:
This function does not support `record` being null, so it should take `record` as a reference.